### PR TITLE
IMU Integration

### PR DESCRIPTION
**Pull request by kylestach**
## Description
This PR contains the beginning of integration of IMU into the control stack. Most of this was tested on-robot before quarantine, and works pretty nicely (still pending some wheel slip issues).

This overhauls:
 - Robot estimator: gives the KF static steady-state gains, which was a little easier to work with and tune
 - Robot controller: reverts to a very simple body velocity controller that attempts to hold zero velocity
 - IMU drivers (ICM 20948): pretty much written from the ground up. Includes accelerometer readings.

### CODE REVIEW

File: robot/lib/Src/drivers/ICM20948.cpp [hunk 1]
```diff
@@ -0,0 +1,189 @@
+#include "drivers/ICM20948.hpp"
```
Comment (JNeiger) [inline robot/lib/Src/drivers/ICM20948.cpp:1]: I'm conflicted. It's really standard when doing these drivers to just have the functionality for everything exposed and let the user do all the config. But having a second wrapper class or placing the config into the module really dirties up the repo. This is probably the best way around that, but it feels wrong from an embedded pov

Comment (JNeiger) [reply robot/lib/Src/drivers/ICM20948.cpp:1]: I haven't checked the actual driver yet, I'll do it another time

**Comment by Arvind-Srinivasan** [comment]
@kylestach any reason this branch has merge conflicts with master?

File: robot/control/Src/modules/FPGAModule.cpp [hunk 2]
```diff
@@ -114,28 +119,23 @@ void FPGAModule::entry() {
...
             motorFeedbackLock->encoders[i] =
-                    static_cast<float>(encDeltas[i]) * (1 / static_cast<float>(ENC_TICK_PER_REV)) * (2 * M_PI / 1) *
-                    (1 / dt);
-        }
-
-        // Convert from adc lsb to amp
-        for (int i = 0; i < 4; i++) {
-            motorFeedbackLock->currents[i] = 0.0f;
+                    static_cast<float>(encDeltas[i]) / ENC_TICK_PER_REV
+                    * 2 * M_PI * kFrequency;
```
[the remark this answers is not in the document]
Comment (kylestach) [reply robot/control/Src/modules/FPGAModule.cpp:138]: Considering we have had issues in the past with that `dt` being zero/very small, I honestly trust the RTOS timer to be consistent more than I trust the timestamp we get off of the FPGA.

File: robot/control/Src/motion-control/RobotEstimator.cpp [hunk 3]
```diff
@@ -21,44 +21,50 @@ RobotEstimator::RobotEstimator(uint32_t dt_us) {
...
-    z(4) = x_hat(2);
-    K << 1.83624980e-01, -2.29806179e-01, -2.29806179e-01, 1.83624980e-01, -2.13457449e-04,
-         3.07632598e-01, 2.76060529e-01, -2.76060529e-01, -3.07632598e-01, -2.99584147e-18,
-         -6.86033585e-02, -6.02908809e-02, -6.02908809e-02, -6.86033585e-02, 2.38627208e-02;
 
-    x_hat += K * 0.01 * y;
+    K <<
+        0.001941, -0.002376, -0.002376, 0.001941, -0.001040,
+        0.001117, 0.001002, -0.001002, -0.001117, 0.000000,
+        -0.004554, -0.004132, -0.004132, -0.004554, 0.152357;
```
[the remark this answers is not in the document]
Comment (kylestach) [reply robot/control/Src/motion-control/RobotEstimator.cpp:49]: Steady-state KF gain from a python script. @tcontis is working on integrating this process directly so we don't have to copy-paste (i.e. a python file generates a .cpp with the matrix definition).

File: robot/control/main.cpp [hunk 4]
```diff
@@ -101,9 +101,24 @@ void createModule(GenericModule *module) {
...
+
+    void entry() override {
+        pin.write(value);
+        value = !value;
+    }
+
+private:
+    DigitalOut pin;
+    bool value = false;
+};
```
[the remark this answers is not in the document]
Comment (kylestach) [reply robot/control/main.cpp:116]: ACK.

File: robot/lib/CMakeLists.txt [hunk 5]
```diff
@@ -9,8 +9,9 @@ add_library(firm-lib
   Src/drivers/ISM43340.cpp
   Src/drivers/KickerBoard.cpp
   Src/drivers/MCP23017.cpp
-  Src/drivers/MPU6050.cpp)
-
+  Src/drivers/MPU6050.cpp
```
Comment (Arvind-Srinivasan) [inline robot/lib/CMakeLists.txt:12]: Can we remove this?

File: robot/control/Src/modules/MotionControlModule.cpp [hunk 6]
```diff
@@ -86,7 +86,6 @@ void MotionControlModule::entry() {
     // Leaving this hear until someone can test, then remove
     // this
     // - Joe Aug 2019
```
Comment (Arvind-Srinivasan) [inline robot/control/Src/modules/MotionControlModule.cpp:88]: Probably should remove

File: robot/lib/Inc/drivers/ICM20948.hpp [hunk 7]
```diff
@@ -13,13 +13,12 @@ class ICM20948 {
...
     double gyro_z();
-    
+
     double accel_x();
 
     double accel_y();
 
-    <template T>
-    std::vector<T> burst_read(uint8_t bank, uint8_t address, uint8_t length);
+    void burst_read(uint8_t bank, uint8_t address, uint8_t* buffer_out, size_t length);
```
Comment (Arvind-Srinivasan) [inline robot/lib/Inc/drivers/ICM20948.hpp:21]: Do we have a test for this?

**Review by Arvind-Srinivasan — APPROVED** [review]
Overall seems solid.

File: robot/lib/Src/drivers/ICM20948.cpp [hunk 8]
```diff
@@ -206,8 +202,9 @@ uint8_t ICM20948::read_register(uint8_t bank, uint8_t address) {
     return data;
 }
 
-<template T>
-std::vector<T> ICM20948::burst_read(uint8_t bank, uint8_t address, uint8_t length) {
+void ICM20948::burst_read(uint8_t bank, uint8_t address, uint8_t* buffer_out, size_t length) {
```
Comment (JNeiger) [inline robot/lib/Src/drivers/ICM20948.cpp:205]: Not sure if intentional, but burst reads aren't being used in the read* (eg accel_x) type calls you expect them to be used in

**Review by JNeiger — APPROVED** [review]

File: robot/lib/CMakeLists.txt (hunk 5 above)
Comment (kylestach) [reply robot/lib/CMakeLists.txt:12]: I'd rather not remove drivers for devices we no longer use. They can still be useful in cases.

**Pull request merged**